File: atcoder.jp/contests/abs/tasks/arc065_a/main.py
from typing import Union

CANDIDATES = ("dream", "dreamer", "erase", "eraser")
REVERSED_CANDIDATES = tuple(s[::-1] for s in CANDIDATES)


def yes_no(b: Union[bool, int]) -> str:
    return "YES" if b else "NO"


def solve(s: str) -> bool:
    while s:
        try:
            prefix = next(p for p in REVERSED_CANDIDATES if s.startswith(p))
        except StopIteration:
            return False
        s = s[len(prefix) :]
    return True


def main() -> None:
    S = input()

    ans = solve(S[::-1])

    print(yes_no(ans))


main()

File: atcoder.jp/contests/abs/tasks/arc065_a/main.ts
import { readFileSync } from "fs";
let inputLinesIndex = 0;
const inputLines = readFileSync("/dev/stdin", "utf8").split("\n");
const input = () => inputLines[inputLinesIndex++];
const print = (...messages: any) => console.log(...messages);

const reverse = (s: string) => s.split("").reverse().join("");
const yesNo = (b: boolean | number) => (b ? "YES" : "NO");

const CANDIDATES = ["dream", "dreamer", "erase", "eraser"];
const REVERSED_CANDIDATES = CANDIDATES.map(reverse);

const solve = (s: string): boolean => {
  while (s) {
    const prefix = REVERSED_CANDIDATES.find((p) => s.startsWith(p));
    if (!prefix) return false;
    s = s.slice(prefix.length);
  }
  return true;
};

const main = () => {
  const S = input();

  const ans = solve(reverse(S));

  print(yesNo(ans));
};

main();
